go-controller/pkg/clustermanager/userdefinednetwork/controller/userdefinednetwork/controller.go

```diff
@@ -0,0 +1,364 @@
+	udnLister := udnInformer.Lister()
+	c := &Controller{
+		nadClient:   nadClient,
+		nadLister:   nadInfomer.Lister(),
+		udnClient:   udnClient,
+		udnLister:   udnLister,
+		nadRenderer: nadRenderer,
+		podInformer: podInformer,
+	}
+	cfg := &controller.ControllerConfig[userdefinednetworkv1.UserDefinedNetwork]{
+		RateLimiter:    workqueue.DefaultControllerRateLimiter(),
+		Reconcile:      c.UserDefinedNetworkReconciler,
+		ObjNeedsUpdate: c.udnNeedUpdate,
+		Threadiness:    1,
+		Informer:       udnInformer.Informer(),
+		Lister:         udnLister.List,
+	}
+	const udnControllerName = "user-defined-network-controller"
+	c.Controller = controller.NewController[userdefinednetworkv1.UserDefinedNetwork](udnControllerName, cfg)
+
+	return c
+}
+
+func (c *Controller) udnNeedUpdate(_, _ *userdefinednetworkv1.UserDefinedNetwork) bool {
+	return true
+}
+
+func (c *Controller) UserDefinedNetworkReconciler(key string) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	udn, err := c.udnLister.UserDefinedNetworks(namespace).Get(name)
+	if err != nil && !kerrors.IsNotFound(err) {
+		return fmt.Errorf("failed to get UserDefinedNetwork %q from cache: %v", key, err)
+	}
+
+	nad, err := c.nadLister.NetworkAttachmentDefinitions(namespace).Get(name)
+	if err != nil && !kerrors.IsNotFound(err) {
+		return fmt.Errorf("failed to get NetworkAttachmetDefinition %q from cache: %v", key, err)
+	}
+
+	udnCopy := udn.DeepCopy()
+	nadCopy := nad.DeepCopy()
+
+	syncErr := c.SyncUserDefinedNetwork(udnCopy, nadCopy)
+
+	updateStatusErr := c.UpdateUserDefinedNetworkStatus(udnCopy, nadCopy, syncErr)
+
+	return errors.Join(syncErr, updateStatusErr)
+}
+
+func (c *Controller) SyncUserDefinedNetwork(udn *userdefinednetworkv1.UserDefinedNetwork, nad *netv1.NetworkAttachmentDefinition) error {
+	if udn == nil {
+		return nil
+	}
+
+	if udn.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(udn, FinalizerUserDefinedNetwork) {
+			controllerutil.AddFinalizer(udn, FinalizerUserDefinedNetwork)
+			patch, err := newFinalizersReplaceJsonPatch(udn.Finalizers)
+			if err != nil {
+				return err
+			}
+			_, err = c.udnClient.K8sV1().UserDefinedNetworks(udn.Namespace).Patch(context.Background(), udn.Name, types.JSONPatchType, patch, metav1.PatchOptions{})
+			if err != nil {
+				return err
+			}
+			klog.Infof("Added finalizer to UserDefinedNetwork [%s/%s]", udn.Namespace, udn.Name)
+			return nil
+		}
+	} else { // udn is being  deleted
+		if controllerutil.ContainsFinalizer(udn, FinalizerUserDefinedNetwork) {
+
+			if nad != nil && controllerutil.ContainsFinalizer(nad, FinalizerUserDefinedNetwork) {
```

Comment:
Since the UDN and NAD object keys is the same, and NAD is created with owner-refernce and a finalizer.
I think there is no scenario it"ll reach this point with a NAD that is not managed by the controller.
Make sense?

Reply:
I think this may happen: as an admin I create NAD with name="test" and I add "user-defined-network-protection" finalizer (for whatever reason). Now I create UDN in that namespace with name="test", which hopefully fails, I realize what the problem is and delete that UDN. Now if we don't check NAD owner reference, we will consider existing NAD to be our own, while it is in reality was not created by our controller and doesn't have an owner reference.

Reply:
Thanks for explaining, done.